File: src/exercise_action_cpp/src/rotate_robot_client.cpp
#include "base_interfaces/action/rotate_robot.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include <cmath>
#include <exception>
#include <string>

using namespace std::chrono_literals;
using base_interfaces::action::RotateRobot;
using std::placeholders::_1;
using std::placeholders::_2;

class RotateRobotClient : public rclcpp::Node {
    friend int main(int argc, char **argv);

public:
    RotateRobotClient()
        : Node("rotate_robot_client_cpp"), is_cancel_(false), cancel_pending_(false), goal_finished_(false),
          cancellation_enabled_(false), cancel_progress_(0.0f) {
        RCLCPP_INFO(this->get_logger(), "动作客户端创建成功!");
        client_ = rclcpp_action::create_client<RotateRobot>(this, "rotate_robot");
        cancel_timer_ = this->create_wall_timer(10ms, std::bind(&RotateRobotClient::process_cancel_request, this));
    }

    bool send_goal(double target_angle, double angular_speed) {
        if (!client_->wait_for_action_server(10s)) {
            RCLCPP_ERROR(this->get_logger(), "服务器连接失败!");
            return false;
        } else {
            RCLCPP_INFO(this->get_logger(), "服务器连接成功!");
        }

        auto goal = RotateRobot::Goal();
        auto options = rclcpp_action::Client<RotateRobot>::SendGoalOptions();

        goal.target_angle = target_angle;
        goal.angular_speed = angular_speed;

        options.goal_response_callback = std::bind(&RotateRobotClient::goal_response_callback, this, _1);
        options.feedback_callback = std::bind(&RotateRobotClient::feedback_callback, this, _1, _2);
        options.result_callback = std::bind(&RotateRobotClient::result_callback, this, _1);

        /*
            std::shared_future<rclcpp_action::ClientGoalHandle<base_interfaces::action::RotateRobot>::SharedPtr> async_send_goal(
                const base_interfaces::action::RotateRobot::Goal &goal,
                const rclcpp_action::Client<base_interfaces::action::RotateRobot>::SendGoalOptions &options
            )
        */
        client_->async_send_goal(goal, options);
        return true;
    }

private:
    /*
        using GoalHandle = ClientGoalHandle<ActionT>;
        using WrappedResult = typename GoalHandle::WrappedResult;
        using GoalResponseCallback = std::function<void (typename GoalHandle::SharedPtr)>;
    */
    void goal_response_callback(rclcpp_action::ClientGoalHandle<RotateRobot>::SharedPtr goal_handle) {
        if (goal_handle) {
            RCLCPP_INFO(this->get_logger(), "目标设置成功!");

            if (cancellation_enabled_ && cancel_progress_ <= 0.0f) {
                pending_goal_handle_ = goal_handle;
                cancel_pending_ = true;
                RCLCPP_INFO(this->get_logger(), "取消进度为0%%, 等待发送取消请求!");
            }
        } else {
            goal_finished_ = true;
            RCLCPP_ERROR(this->get_logger(), "目标设置失败!");
            stop_client();
        }
    }

    /*
        std::function<void (typename ClientGoalHandle<ActionT>::SharedPtr, const std::shared_ptr<const Feedback>)>;
    */
    void feedback_callback(
        rclcpp_action::ClientGoalHandle<RotateRobot>::SharedPtr goal_handle,
        const std::shared_ptr<const RotateRobot::Feedback> feedback
    ) {
        RCLCPP_INFO(
            this->get_logger(),
            "连续反馈中: [current_angle = %lf, remaining_angle = %lf, progress = %d%%]",
            feedback->current_angle,
            feedback->remaining_angle,
            static_cast<int>(std::round(feedback->progress * 100))
        );

        if (
            cancellation_enabled_ && !is_cancel_ && !cancel_pending_ && !goal_finished_ &&
            feedback->progress >= cancel_progress_ && feedback->progress < 1.0f
        ) {
            pending_goal_handle_ = goal_handle;
            cancel_pending_ = true;
            RCLCPP_INFO(this->get_logger(), "已达到取消进度, 等待发送取消请求!");
        }
    }

    void process_cancel_request() {
        if (goal_finished_) {
            pending_goal_handle_.reset();
            cancel_pending_ = false;
            return;
        }

        if (!cancel_pending_ || !pending_goal_handle_) {
            return;
        }

        auto goal_handle = pending_goal_handle_;
        pending_goal_handle_.reset();
        cancel_pending_ = false;
        is_cancel_ = true;

        try {
            /*
            std::shared_future<action_msgs::srv::CancelGoal_Response_<std::allocator<void>>::SharedPtr>
            async_cancel_goal(
                std::shared_ptr<rclcpp_action::ClientGoalHandle<base_interfaces::action::RotateRobot>> goal_handle,
                std::function<void (std::shared_ptr<action_msgs::srv::CancelGoal_Response>)> cancel_callback
            )
            */
            client_->async_cancel_goal(
                goal_handle,
                std::bind(&RotateRobotClient::cancel_response_callback, this, _1)
            );
            RCLCPP_INFO(this->get_logger(), "已发出任务取消请求!");
        } catch (const std::exception &exception) {
            is_cancel_ = false;
            RCLCPP_ERROR(this->get_logger(), "发送取消请求失败: %s", exception.what());
        }
    }

    void cancel_response_callback(rclcpp_action::Client<RotateRobot>::CancelResponse::SharedPtr response) {
        if (response->goals_canceling.empty()) {
            RCLCPP_WARN(this->get_logger(), "取消请求未被服务端接受!");
        } else {
            RCLCPP_INFO(this->get_logger(), "服务端已接受取消请求!");
        }
    }

    /*
        std::function<void (const WrappedResult & result)>;
    */
    void result_callback(const rclcpp_action::ClientGoalHandle<RotateRobot>::WrappedResult &result) {
        goal_finished_ = true;
        cancel_pending_ = false;
        pending_goal_handle_.reset();

        switch (result.code) {
            case rclcpp_action::ResultCode::SUCCEEDED:
                RCLCPP_INFO(this->get_logger(), "任务执行完成!");
                break;

            case rclcpp_action::ResultCode::ABORTED:
                RCLCPP_WARN(this->get_logger(), "任务被中断!");
                break;

            case rclcpp_action::ResultCode::CANCELED:
                RCLCPP_WARN(this->get_logger(), "任务被取消!");
                break;

            default:
                RCLCPP_WARN(this->get_logger(), "收到未知的任务结果!");
                stop_client();
                return;
        }

        if (result.result) {
            RCLCPP_INFO(
                this->get_logger(),
                "[success = %s, message = %s, final_angle = %lf, elapsed_time = %lf]",
                result.result->success ? "true" : "false",
                result.result->message.c_str(),
                result.result->final_angle,
                result.result->elapsed_time
            );
        }

        stop_client();
    }

    void stop_client() {
        if (rclcpp::ok()) {
            rclcpp::shutdown();
        }
    }

    bool is_cancel_;
    bool cancel_pending_;
    bool goal_finished_;
    bool cancellation_enabled_;
    float cancel_progress_;

    rclcpp_action::ClientGoalHandle<RotateRobot>::SharedPtr pending_goal_handle_;
    rclcpp_action::Client<RotateRobot>::SharedPtr client_;
    rclcpp::TimerBase::SharedPtr cancel_timer_;
};

bool parse_finite_number(const char *text, const char *parameter_name, double &value) {
    const std::string input(text);
    std::size_t parsed_characters = 0;

    try {
        value = std::stod(input, &parsed_characters);
    } catch (const std::exception &) {
        RCLCPP_ERROR(
            rclcpp::get_logger("rotate_robot_client_cpp"),
            "%s必须是有效数值: %s",
            parameter_name,
            text
        );
        return false;
    }

    if (parsed_characters != input.size() || !std::isfinite(value)) {
        RCLCPP_ERROR(
            rclcpp::get_logger("rotate_robot_client_cpp"),
            "%s必须是有效的有限数值: %s",
            parameter_name,
            text
        );
        return false;
    }

    return true;
}

int main(int argc, char **argv) {
    if (argc != 3 && argc != 4) {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "输入参数不正确!");
        return 1;
    }

    rclcpp::init(argc, argv);

    double target_angle = 0.0;
    double angular_speed = 0.0;
    double cancel_percentage = 0.0;

    if (
        !parse_finite_number(argv[1], "目标角度", target_angle) ||
        !parse_finite_number(argv[2], "角速度", angular_speed)
    ) {
        rclcpp::shutdown();
        return 1;
    }

    if (argc == 4) {
        if (!parse_finite_number(argv[3], "取消进度", cancel_percentage)) {
            rclcpp::shutdown();
            return 1;
        }

        if (cancel_percentage < 0.0 || cancel_percentage >= 100.0) {
            RCLCPP_ERROR(rclcpp::get_logger("rotate_robot_client_cpp"), "取消进度必须在[0, 100)范围内!");
            rclcpp::shutdown();
            return 1;
        }
    }

    auto client = std::make_shared<RotateRobotClient>();

    if (argc == 4) {
        client->cancellation_enabled_ = true;
        client->cancel_progress_ = static_cast<float>(cancel_percentage * 0.01);
    }

    if (!client->send_goal(target_angle, angular_speed)) {
        rclcpp::shutdown();
        return 1;
    }

    rclcpp::spin(client);

    if (rclcpp::ok()) {
        rclcpp::shutdown();
    }

    return 0;
}

File: src/exercise_action_py/exercise_action_py/rotate_robot_client.py
import math
import sys
from typing import Optional, cast

import rclpy
from action_msgs.msg import GoalStatus
from action_msgs.srv import CancelGoal
from base_interfaces.action import RotateRobot
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle
from rclpy.logging import get_logger
from rclpy.node import Node
from rclpy.task import Future


class RotateRobotClient(Node):
    def __init__(self):
        super().__init__('rotate_robot_client_py')
        self.get_logger().info('动作客户端创建成功!(Python)')

        self.client_ = ActionClient(self, RotateRobot, 'rotate_robot')
        self.timer_ = self.create_timer(0.01, self.timer_callback)

        self.cancel_start_ = False
        self.cancel_enable_ = False
        self.cancel_progress_ = 0.0
        self.finished_ = False
        self.exit_code_ = 0
        self.goal_handle_: Optional[ClientGoalHandle] = None

    def send_goal(self, target_angle: float, angular_speed: float) -> bool:
        if not self.client_.wait_for_server(10.0):
            self.get_logger().error('服务器连接超时, 客户端退出!')
            return False

        goal = RotateRobot.Goal()
        goal.target_angle = target_angle
        goal.angular_speed = angular_speed

        self.get_logger().info(
            '设置的目标为: '
            f'[target_angle = {goal.target_angle}, '
            f'angular_speed = {goal.angular_speed}]'
        )

        self.goal_future_ = self.client_.send_goal_async(
            goal,
            self.feedback_callback
        )
        self.goal_future_.add_done_callback(self.goal_response_callback)
        return True

    def goal_response_callback(self, future: Future):
        try:
            self.goal_handle_ = cast(ClientGoalHandle, future.result())
        except Exception as exception:
            self.get_logger().error(f'获取目标响应失败: {exception}')
            self.finish(1)
            return

        if self.goal_handle_ is None or not self.goal_handle_.accepted:
            self.get_logger().error('目标被拒绝!')
            self.finish(1)
            return

        self.get_logger().info('目标被接受!')

        if self.cancel_enable_ and self.cancel_progress_ <= 0.0:
            self.cancel_start_ = True
            self.get_logger().info('取消进度为0%, 等待发送取消请求!')

        self.result_future_ = self.goal_handle_.get_result_async()
        self.result_future_.add_done_callback(self.get_result_callback)

    def feedback_callback(
        self,
        feedback_msg: RotateRobot.Impl.FeedbackMessage
    ):
        feedback = feedback_msg.feedback

        self.get_logger().info('====================')
        self.get_logger().info('连续接收反馈中...')
        self.get_logger().info(
            f'current_angle = {feedback.current_angle}'
        )
        self.get_logger().info(
            f'remaining_angle = {feedback.remaining_angle}'
        )
        self.get_logger().info(
            f'progress = {round(feedback.progress * 100)}%'
        )
        self.get_logger().info('====================')

        if not self.cancel_enable_ or self.finished_:
            return

        if (
            feedback.progress >= self.cancel_progress_
            and feedback.progress < 1.0
        ):
            self.cancel_start_ = True

    def get_result_callback(self, future: Future):
        try:
            wrapped_result = future.result()
        except Exception as exception:
            self.get_logger().error(f'获取任务结果失败: {exception}')
            self.finish(1)
            return

        status = wrapped_result.status
        result = wrapped_result.result

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info('任务执行完成!')
            exit_code = 0
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warning('任务已取消!')
            exit_code = 0
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error('任务被中断!')
            exit_code = 1
        else:
            self.get_logger().error(f'未知的任务状态: {status}')
            exit_code = 1

        self.get_logger().info('====================')
        self.get_logger().info('任务处理结果如下')
        self.get_logger().info(f'success = {result.success}')
        self.get_logger().info(f'message = {result.message}')
        self.get_logger().info(f'final_angle = {result.final_angle}')
        self.get_logger().info(f'elapsed_time = {result.elapsed_time}')
        self.get_logger().info('====================')

        self.finish(exit_code)

    def cancel_response_callback(self, future: Future):
        try:
            response = cast(CancelGoal.Response, future.result())
        except Exception as exception:
            self.get_logger().error(f'取消请求处理失败: {exception}')
            return

        if response.goals_canceling:
            self.get_logger().info('服务端已接受取消请求!')
        else:
            self.get_logger().warning(
                '服务端未接受取消请求, '
                f'return_code = {response.return_code}'
            )

    def timer_callback(self):
        if not self.cancel_start_ or self.finished_:
            return

        self.cancel_start_ = False
        self.cancel_enable_ = False

        if self.goal_handle_ is None:
            return

        self.get_logger().info('已发送取消请求, 等待任务取消!')

        self.cancel_future_ = self.goal_handle_.cancel_goal_async()
        self.cancel_future_.add_done_callback(
            self.cancel_response_callback
        )

    def finish(self, exit_code: int):
        self.finished_ = True
        self.cancel_start_ = False
        self.cancel_enable_ = False
        self.exit_code_ = exit_code

        if rclpy.ok():
            rclpy.shutdown()


def parse_finite_number(
    text: str,
    parameter_name: str
) -> Optional[float]:
    try:
        value = float(text)
    except ValueError:
        get_logger('rotate_robot_client_py').error(
            f'{parameter_name}必须是有效数值: {text}'
        )
        return None

    if not math.isfinite(value):
        get_logger('rotate_robot_client_py').error(
            f'{parameter_name}必须是有限数值: {text}'
        )
        return None

    return value


def main():
    if len(sys.argv) not in (3, 4):
        get_logger('rotate_robot_client_py').error('提交参数数目不合法!')
        return 1

    target_angle = parse_finite_number(sys.argv[1], '目标角度')
    angular_speed = parse_finite_number(sys.argv[2], '角速度')

    if target_angle is None or angular_speed is None:
        return 1

    cancel_percentage = None
    if len(sys.argv) == 4:
        cancel_percentage = parse_finite_number(sys.argv[3], '取消进度')
        if cancel_percentage is None:
            return 1

        if cancel_percentage < 0.0 or cancel_percentage >= 100.0:
            get_logger('rotate_robot_client_py').error(
                '取消进度必须在[0, 100)范围内!'
            )
            return 1

    rclpy.init()
    client = RotateRobotClient()

    if cancel_percentage is not None:
        client.cancel_enable_ = True
        client.cancel_progress_ = cancel_percentage * 0.01

    if not client.send_goal(target_angle, angular_speed):
        client.exit_code_ = 1
    else:
        try:
            rclpy.spin(client)
        except KeyboardInterrupt:
            pass

    client.destroy_node()

    if rclpy.ok():
        rclpy.shutdown()

    return client.exit_code_


if __name__ == '__main__':
    main()
